components/canvas/canvas_data.rs

```diff
@@ -671,21 +671,25 @@ impl<DrawTarget: GenericDrawTarget> CanvasData<DrawTarget> {
         rect: &Rect<f32>,
         shadow_options: ShadowOptions,
         composition_options: CompositionOptions,
-        transform: Transform2D<f32>,
+        transform: Transform2D<f64>,
         draw_shadow_source: F,
     ) where
-        F: FnOnce(&mut DrawTarget, Transform2D<f32>),
+        F: FnOnce(&mut DrawTarget, Transform2D<f64>),
     {
-        let shadow_src_rect = transform.outer_transformed_rect(rect);
-        let mut new_draw_target = self.create_draw_target_for_shadow(&shadow_src_rect);
+        let shadow_src_rect = transform.outer_transformed_rect(&rect.cast());
+        // Because this comes from the rect on f32 precision, casting it down should be ok.
+        let mut new_draw_target = self.create_draw_target_for_shadow(&shadow_src_rect.cast());
```

Comment:
Usually it is better to do computation in bigger precision then casting it back to lower precision, although I suspect all stuff in 2D canvas should be f64.